### PR TITLE
[auto-materialize] refactor reasons

### PR DESCRIPTION
**Pull request by OwenKephart**
## Summary & Motivation

Now that we have 3 different states that a condition may belong to, the "separate enum and namedtuple and field per state" thing was getting pretty ugly, as we were using the result ("MATERIALIZE", "SKIP", or "DISCARD") as a top level concept which determined class names.

This refactors thing to make everything built on top of these results, rather than vice versa. So we have an enum indicating the type of outputs we can get for a decision, then an enum indicating the types of conditions that can lead to those results, then finally a NamedTuple that can bundle this condition with extra metadata.

In general, this results in way less code than the alternative solution, and also makes it easier to add new results / conditions.

### CODE REVIEW

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_reason.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
+from enum import Enum
+from typing import NamedTuple
+
+
+class AutoMaterializeResult(Enum):
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/auto_materialize_reason.py:5]: What do you think about calling this `AutoMaterializeResultType`? I did a grep for Enum in our codebase, and it seems like they basically all end in either "Type" or "Status".

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_reason.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+
+class AutoMaterializeCondition(Enum):
+    """Represents the set of conditions that can impact auto-materialization for an asset."""
+
+    def __init__(self, id: str, result: AutoMaterializeResult):
+        self.id = id
+        self.result = result
+
+    # Materialization Reasons
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/auto_materialize_reason.py:27]: Is this use of "Reason" congruous with the name `AutoMaterializeConditionReason`?

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_reason.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    MISSING = ("MISSING", AutoMaterializeResult.MATERIALIZE)
+
+    # Skip Reasons
+    PARENT_OUTDATED = ("PARENT_OUTDATED", AutoMaterializeResult.SKIP)
+
+    # Discard Reasons
+
+
+class AutoMaterializeConditionReason(NamedTuple):
+    """Denotes the reason that the auto-materialize logic decided that an asset should be materialized.
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/auto_materialize_reason.py:40]: I'm finding this a little difficult to parse. Is this class only relevant in the case that the `AutoMaterializeResult` of the `AutoMaterializeCondition` is MATERIALIZE? From the name, I would expect it to apply to any `AutoMaterializeCondition`.

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_reason.py (hunk 1 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_reason.py:5]: yeah makes sense

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_reason.py (hunk 2 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_reason.py:27]: should be `"Materialize Conditions"`

File: python_modules/dagster/dagster/_core/definitions/auto_materialize_reason.py (hunk 3 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_reason.py:40]: I just forgot to update this docstring -- this is just a wrapper class that allows (in the future) more specific information to be attached to the underlying condition

Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/auto_materialize_reason.py:40]: e.g. the "condition" is "DOWNSTREAM_FRESHNESS", and the specific data is "Assets X, Y, and Z needed fresher data", or the "condition" is "PARENT_NEVER_MATERIALIZED", and the specific data is "Upstream asset A has never been materialized"

**Comment by OwenKephart** [comment]
I think "reason" is bad here, as "AutoMaterializeReason" implies "the reason that this asset was auto materialized", whereas a) some of these reasons imply that the asset will NOT be auto materialized and b) multiple of these objects may apply to each asset-partition.

Some alternatives:

AutoMaterializeFactor
AutoMaterializeCriterion
AutoMaterializeInput
AutoMaterializeCondition

Went with Condition for now.

**Review by sryza — APPROVED** [review]
Nice

Maybe `AutoMaterializeDecisionCondition`? But is more wordy and might not be better.

**Pull request merged**